### DISCUSSION TITLE
Dev for Dev 컴포넌트 분석 (common)

### DISCUSSION THREAD

**Discussion by KingNono1030** [DevFDev/frontend#36] [Announcements]
# Common Component
- #13 
> common은 아토믹한 컴포넌트로, 도메인 맥락이 없는 가장 작은 단위의 컴포넌트

## 1. Button : 유저가 클릭할 수 있는 모든 요소
<img width="309" alt="image" src="https://github.com/user-attachments/assets/811c691e-c335-4876-8afb-0b8046bb8a5c">
<img width="341" alt="image" src="https://github.com/user-attachments/assets/21ee93de-5349-421c-b2f6-7f2961efa6ab">
<img width="433" alt="image" src="https://github.com/user-attachments/assets/39afec2f-b19d-40fd-a019-a5170d42c659">
<img width="338" alt="image" src="https://github.com/user-attachments/assets/3c796a8e-46e7-4bde-8910-9382c62e02a7">

```
components
└── common
    └── Button
        ├── Clickable.tsx  // 버튼과 링크의 공통 스타일링 및 기능
        ├── Button.tsx     // 버튼 컴포넌트
        └── Link.tsx       // 링크 컴포넌트
```
> 버튼 vs Link 고민
- 디자인의 경우 특정 페이지로 이동 시키는 Link 와 Button 에서 공통점이 있어
- 따라서 Clickable 이라는 공통의 스타일링 컴포넌트를 Button 과 Link 를 공유하게 하고자 함
- 얘는 스타일과 관련된 props 를 받는데, variant, size, additional effect (hover, active, disabled), icon, iconLeft 등이 올 것 같음

> 구현 예시
```tsx
// Clickable: 스타일과 기본 인터랙션을 관리하는 공통 스타일 컴포넌트
import clsx from 'clsx';

function Clickable({ children, variant, size, ...props }) {
  const style = clsx('clickable', variant, size);
  return <span className={style} {...props}>{children}</span>;
}

// Button: Clickable을 확장한 기본 버튼 컴포넌트
function Button({ children, ...props }) {
  return <button {...props}><Clickable {...props}>{children}</Clickable></button>;
}

// Link: 페이지 이동을 위한 링크 컴포넌트
function Link({ children, to, ...props }) {
  return <NextLink href={to} {...props}><Clickable {...props}>{children}</Clickable></NextLink>;
}

```
## 2. Chip: UI 기능만 하는, 특정 라벨을 감싸는 둥근 직사각형 꼴의 블록 엘리먼트
<img width="669" alt="image" src="https://github.com/user-attachments/assets/8a79f683-f4cc-4a54-8732-e211bf2f612f">

```
components
└── common
    └── Chip
        ├── Chip.tsx         // 기본 Chip 컴포넌트
        └── TagChip.tsx      // 태그 관리 기능이 포함된 TagChip 컴포넌트
```

@KingNono1030 
- 칩은 게시글 등의 성격을 보여주는 라벨의 역할. 구인글에는 스터디, 프로젝트, 멘토링 / 커뮤니티 글에는 기술, 커리어. 기타 등이 해당. 이들에게는 각각의 고유 색깔이 적용되고 있고
- 그 다음에는 특정한 분류를 위한 칩. 스타일 면은 동일하나, 주제가 포지션, 기술 스택 등이 해당이 됨
> 기능이 단순하지만, 도메인 별로 분류를 할 수 있을지 고민
> Chip.Position, Chip.
@yongb2n @yellowjang 
>> 과한 분류, 복잡성을 오히려 증가시킴

## 2-1. TagChip: 닫기 버튼이 포함됨
<img width="536" alt="image" src="https://github.com/user-attachments/assets/8b5e850f-1757-441f-bda8-e40045a1df71">

- 사용처: 게시글 입력하는 폼의 태그 인풋 입력시, 하단에 위치
- TagChip은 사용자가 생성한 태그와 관련된 컴포넌트로, 닫기 버튼을 통해 제거 가능한 요소. 반면, Chip은 정보 표시가 주된 목적이므로, 두 컴포넌트를 분리하는게 좋을듯.
- 스타일도 다름

## 4. Input
```
components
└── common
    └── input
        ├── BaseInput.tsx        // 공통 스타일링 제공
        ├── TextInput.tsx        // 기본 텍스트 입력 필드 (기존 InputField)
        ├── PasswordInput.tsx    // 비밀번호 입력 필드
        ├── TagInput.tsx         // 태그 입력 필드
        ├── FileInput.tsx       // 파일 입력 필드
        ├── RadioInput.tsx       // 라디오 버튼 입력 필드
        ├── TextAreaAutoResized.tsx       // 텍스트 애리아 (입력값 추가시 자동으로 사이즈 조정)
        └── CheckboxInput.tsx    // 체크박스 입력 필드
```
## 4-1. Form
```
components
└── shared
    └── Form.tsx            // Form 컴포넌트 본체
```

> Input common components 활용 예시
```
const Input = ({ type, label, ...props }) :JSX.Element => {
  switch (type) {
    case 'text':
      return <TextInput {...props} label={label} />;
    case 'password':
      return <PasswordInput {...props} label={label} />;
    case 'radio':
      return <Radio {...props} label={label} />;
    case 'checkbox':
      return <Checkbox {...props} label={label} />;
    default:
      return new Error('잘못된 input type 제공');
  }
};
```
## 4-2. Switch 컴포넌트

![image](https://github.com/user-attachments/assets/82371500-55d6-4330-85fc-16c83f6b5d2d)
```tsx
    <label className={`switch ${disabled ? 'switch--disabled' : ''}`}>
      <input
        type="checkbox"
        checked={checked}
        onChange={onChange}
        disabled={disabled}
        className="switch__input"
      />
      <span className="switch__slider" />
    </label>
  );
```

## 4-3 Label 컴포넌트
![image](https://github.com/user-attachments/assets/e213d2d5-17f1-4052-90a6-bd14cad6ff5d)

```
components
└── common
    └── label
        ├── Label.tsx       // Label 컴포넌트
        └── Label.test.tsx // Label test 파일 (옵션)
```
> 구현 에시
```tsx
  <label htmlFor={htmlFor} className="label" required={required} {...props}>
    {text}
    {required && <span>*</span>}
  </label>
```
```tsx
Form.Label = Label;
```

## 5. Logo
![image](https://github.com/user-attachments/assets/0f4823d4-5c75-4449-90eb-9f681d5f693c)
- 헤더 로고와 푸터 로고에 따라 다르니까, variant, 또는 position='header / footer' 로 분기처리

## 6. Box
![image](https://github.com/user-attachments/assets/846af695-6aba-4536-a89d-7b2aed090e62)
- Padding, margin, display 등 레이아웃 스타일을 전문으로 하는 Box

## 7. Modal
![image](https://github.com/user-attachments/assets/71462f3d-3c8f-41dd-bc63-7767b476722b)
```
components
└── common
    └── Modal
        ├── Modal.tsx           // 기본 모달 컴포넌트
        └── ModalPortal.tsx      // 포탈을 통한 위치 관리
```

## 8.Text
![image](https://github.com/user-attachments/assets/dd9a8834-f5e9-4acc-b197-20ff52207ed8)

- 스타일 가이드의 텍스트 스타일에 따라 sub 컴포넌트로 구분
- Text.Heading
- Text.Body
- Text.Title
- Text.Caption
> 어느정도 공통으로 가지는 font weight, 폰트 스타일이 있음

## 9. Avatar
![image](https://github.com/user-attachments/assets/beae21b5-e406-44c6-8a0a-f204df5f05e8)

- size props 받음
- image 안 받으면 기본 이미지 렌더

## 10.Progress bar
![image](https://github.com/user-attachments/assets/2fd3354c-ad70-4375-9189-972e4bae237c)
- progress: number 를 props 로 받아서 파란 막대 div 의 길이 변경

**Comment by yellowjang** [comment]
input 은 그러면 base input 을 import 해서 가공해서 사용하는 건가요 ??

**Reply by KingNono1030** [reply]
> input 은 그러면 base input 을 import 해서 가공해서 사용하는 건가요 ??

네 맞습니다 ! 텍스트 필드가 있는 input -> passowd, tag 의 경우 input 을 import 하면 됩니다.
baseInput 에서 그냥 input 으로 이름 수정했습니다.

최신 내용은 이슈에 먼저 작성을 해서, 컴포넌트 분석 문서는 다시 정리해서 업로드할 예정입니다 !
- #43

**Comment by yongb2n** [comment]
좋습니다!
근데 각 컴포넌트의 기본적인 파일들의 이름은 `index.tsx`로 관리하는 건 어떨까요?

Chip 컴포넌트를 예시로,

```
components
└── common
    └── Chip
        ├── index.tsx         // 기본 Chip 컴포넌트
        └── TagChip.tsx      // 태그 관리 기능이 포함된 TagChip 컴포넌트
